Framework/Core/src/DataDescriptorQueryBuilder.cxx

```diff
@@ -217,28 +246,29 @@ std::vector<InputSpec> DataDescriptorQueryBuilder::parse(char const* config)
           nodes.push_back(DescriptionValueMatcher{*currentDescription});
         } else if (*next == '?' && assignLastStringMatch("description", 16, currentDescription, IN_BEGIN_ATTRIBUTES)) {
           nodes.push_back(DescriptionValueMatcher{*currentDescription});
-        } else if (*next == '\0' && assignLastStringMatch("description", 16, currentDescription, IN_BEGIN_ATTRIBUTES)) {
+        } else if (*next == '\0' && assignLastStringMatch("description", 16, currentDescription, IN_END_MATCHER)) {
           nodes.push_back(DescriptionValueMatcher{*currentDescription});
         } else {
           error("description needs to be between 1 and 16 char long");
         }
       } break;
       case IN_BEGIN_SUBSPEC: {
-        pushState(IN_END_SUBSPEC);
+        checkModifier(IN_END_SUBSPEC, "!");
         token(IN_NUMBER, ";%?");
       } break;
       case IN_END_SUBSPEC: {
         if (*next == '%' && assignLastNumericMatch("subspec", currentSubSpec, IN_BEGIN_TIMEMODULO)) {
-          nodes.push_back(SubSpecificationTypeValueMatcher{*currentSubSpec});
         } else if (*next == '?' && assignLastNumericMatch("subspec", currentSubSpec, IN_BEGIN_ATTRIBUTES)) {
-          nodes.push_back(SubSpecificationTypeValueMatcher{*currentSubSpec});
         } else if (*next == ';' && assignLastNumericMatch("subspec", currentSubSpec, IN_END_MATCHER)) {
-          nodes.push_back(SubSpecificationTypeValueMatcher{*currentSubSpec});
         } else if (*next == '\0' && assignLastNumericMatch("subspec", currentSubSpec, IN_END_MATCHER)) {
-          nodes.push_back(SubSpecificationTypeValueMatcher{*currentSubSpec});
         } else {
           error("Expected a number");
+          break;
         }
+        auto backup = states.back();
+        states.pop_back();
+        pushMatcher(SubSpecificationTypeValueMatcher{*currentSubSpec});
+        states.push_back(backup);
```

Comment:
I'm checking a better solution here, we need to check the state before the one just added in `assignLastNumericMatch`. Have not yet disentangled because the method is in itself checking if the given state can be added.